pilot/pkg/leaderelection/leaderelection.go

```diff
@@ -0,0 +1,115 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package leaderelection
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	coreV1 "k8s.io/api/core/v1"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+	"k8s.io/client-go/tools/record"
+
+	"istio.io/pkg/log"
+)
+
+const (
+	electionId = "istio-leader"
+)
+
+type LeaderElection struct {
+	namespace string
+	name      string
+	runFns    []func(stop <-chan struct{})
+	client    kubernetes.Interface
+}
+
+// Run will start leader election, calling all runFns when we become the leader.
+func (l *LeaderElection) Run(stop <-chan struct{}, onComplete func()) error {
+	le, err := l.create()
+	if err != nil {
+		return fmt.Errorf("failed to create leader election: %v", err)
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		le.Run(ctx)
+		// Register that we have finished leader election
+		// In practice, this typically means we are exiting the process, and should drop our lock
+		if onComplete != nil {
+			onComplete()
+		}
+	}()
+	go func() {
+		<-stop
+		cancel()
+	}()
+	return nil
+}
+
+func (l *LeaderElection) create() (*leaderelection.LeaderElector, error) {
+	callbacks := leaderelection.LeaderCallbacks{
+		OnStartedLeading: func(ctx context.Context) {
+			for _, f := range l.runFns {
+				go f(ctx.Done())
+			}
+		},
+		OnStoppedLeading: func() {
+			log.Infof("leader election lock lost")
+
+		},
+	}
+	broadcaster := record.NewBroadcaster()
+	hostname, _ := os.Hostname()
+	recorder := broadcaster.NewRecorder(scheme.Scheme, coreV1.EventSource{
+		Component: "istio-leader-elector",
+		Host:      hostname,
+	})
+	lock := resourcelock.ConfigMapLock{
+		ConfigMapMeta: metaV1.ObjectMeta{Namespace: l.namespace, Name: electionId},
+		Client:        l.client.CoreV1(),
+		LockConfig: resourcelock.ResourceLockConfig{
+			Identity:      l.name,
+			EventRecorder: recorder,
+		},
+	}
+	ttl := 30 * time.Second
+	return leaderelection.NewLeaderElector(leaderelection.LeaderElectionConfig{
+		Lock:          &lock,
+		LeaseDuration: ttl,
+		RenewDeadline: ttl / 2,
+		RetryPeriod:   ttl / 4,
+		Callbacks:     callbacks,
+		// When Pilot exits, the lease will be dropped. This is more likely to lead to a case where
+		// to instances are both considered the leaders. As such, if this is intended to be use for mission-critical
+		// usages (rather than avoiding duplication of work), this may need to be re-evaluated.
+		ReleaseOnCancel: true,
+	})
+}
+
+// AddRunFunction registers a function to run when we are the leader. These will be run asynchronously.
+// To avoid running when not a leader, functions should respect the stop channel.
+func (l *LeaderElection) AddRunFunction(f func(stop <-chan struct{})) {
```

Comment:
@howardjohn there are a few options here:

1. lock access to the `runFns` array wherever it's used
2. Use a separate builder so that the API makes it clear that `runFns` can not be mutated after start.
3. Use a separate `Config` object that you pass to `New...` and this is the thing that you build up before you create the elector.

Personally, I think my preference is 3, it's probably the simplest since you could just have a simple struct with no additional methods, like a builder would require. It also makes the New method clean.

Also, this method should probably be renamed to something like `RunWhenLeader`.